### PR TITLE
feat: add unit test generation

### PR DESCRIPTION
**Pull request by charankamarapu**
Keploy is known to be a end-to-end/integration testing tool. We have decided to add unit test generation along these other testing. 

## Languages Supported
1. JavaScript
2. Golang

Other language may also be supported, but there are yet not tested up on. The Auto detect feature is supported for above languages only.

## Prerequisites
1. Unit Test Report Format - The report format currently we support is Cobertura 
example settings to create Cobertura report: 
    - [javascript](https://github.com/keploy/samples-typescript/blob/f2afb859fac7b25eebe382119db99a3772af6095/express-mongoose/package.json#L36)
    -  Golang
        - Install libraries 
            ```
            go install github.com/axw/gocov/gocov@v1.1.0
            go install github.com/AlekSi/gocov-xml@v1.1.0
            ```
        - test Command 
          ```
          go test -v ./... -coverprofile=coverage.out && gocov convert coverage.out | gocov-xml > coverage.xml
          ```

3. AI model - OpenAI, litellm 
    - User can use either the OpenAI's gpt-4.o by default directly or if the user wants to use other llms user can setup [litellm](https://github.com/BerriAI/litellm?tab=readme-ov-file#quick-start-proxy---cli) and use their model. 
    - To use any model API key is necessary. So user needs to set the env variable `export API_KEY=xxxx `

## Usage

cmd - `keploy gen`
Flags : 
     - `sourceFilePath` - path of the file for which user wants to generate tests
     -  `testFilePath` - path of the file where the tests will be generated
     -  `coverageReportPath`- path of the report where it will be generated
     -  `testCommand` (req) - unit test command to verify if the AI generated tests are working or not
     -  `coverageFormat` - currently we only Cobertura support in future we may add other types too. (default Cobertura) 
     - `expectedCoverage` - Coverage you want to achieve (default 100%) 
     -  `maxIterations` - Number iterations to be done to AI for better test generation (default 5)
     -  `testDir` - directory in which tests needed to written.
     -  `litellmUrl` - base url for litellm proxy
     -  `model` - model to be used in litellm/chatgpt (default gpt-4.o)

## Auto detect feature 

Generally user will be needed to provide the sourceFilePath and testFilePath and tests will be generated for only that file. But feature has been added to identify source files and test files. The simple command would now look like

cmd - `keploy gen --testComamnd="npm test" --testDir="test"` (these two flags are compulsory for auto detect feature to work)

## Usage examples: 

To use this tool please follow the below steps:

1. Installation of keploy tool
    - Currently (PR stage) 
       clone keploy and checkout to unit-test-generation and build 
        ```
        git clone https://github.com/keploy/keploy.git
        cd keploy 
        git checkout unit-test-generation
        go build -o keploy && mv ./keploy /usr/local/bin
        ```
    - (After merging)
       follow - https://keploy.io/docs/server/installation/

3. ### javascript

    - Setup [express-mongoose](https://github.com/keploy/samples-typescript/tree/main/express-mongoose)
    - export API_KEY=xxxx (it needs 4.o access) 
    - run the command to generate tests in the express-mongoose (please replace the paths with your username and directories)

    ```
    keploy gen --sourceFilePath="/Users/charankamarapu/unit-test/samples-typescript/express-mongoose/src/routes/routes.js" --testFilePath="/Users/charankamarapu/unit-test/samples-typescript/express-mongoose/test/routes.test.js" --testCommand="npm test" --coverageReportPath="/Users/charankamarapu/unit-test/samples-typescript/express-mongoose/coverage/cobertura-coverage.xml"
    ```

      - If user wants to generate tests for all files run the below command

    ```
     keploy gen --testCommand="npm test" --coverageReportPath="/Users/charankamarapu/unit-test/samples-typescript/express-mongoose/coverage/cobertura-coverage.xml" --testDir="test"
    ```
  
    

    and the tests will be generated in 2-3 mins and coverage will be increased

4. ### golang 

    - Setup [gin-redis](https://github.com/keploy/samples-go/tree/main/gin-redis)
    - export API_KEY=xxxx (it needs 4.o access) 
    - run the command to generate tests in the gin-redis (please replace the paths with your username and directories)
       
      ```
      keploy gen --sourceFilePath="/Users/charankamarapu/unit-test/samples-go/gin-redis/helpers/token/token.go" --testFilePath="/Users/charankamarapu/unit-test/samples-go/gin-redis/helpers/token/token_test.go" --testCommand="go test -v ./... -coverprofile=coverage.out && gocov convert coverage.out | gocov-xml > coverage.xml" --coverageReportPath="/Users/charankamarapu/unit-test/samples-go/gin-redis/coverage.xml"
      ```

    -  If user wants to generate tests for all files run the below command

      ```
      keploy gen  --testDir="." --testCommand="go test -v ./... -coverprofile=coverage.out && gocov convert coverage.out | gocov-xml > coverage.xml" --coverageReportPath="/Users/charankamarapu/unit-test/samples-go/gin-redis/coverage.xml"
      ```

### CODE REVIEW

**Comment by slayerjain** [comment]
@charankamarapu This looks awesome. I tried running and was able to increase coverage in the sample project from 34% or 49% How much were you able to achieve?

**Comment by slayerjain** [comment]
I ran again, and I got 58%. Maybe GPT4o got warmed up. 

```
🐰 Keploy: 2024-05-31T22:18:39+05:30 	INFO	Total token used count for LLM model gpt-4o: 1472
🐰 Keploy: 2024-05-31T22:18:39+05:30 	INFO	Running test with the following command: 'npm test'
🐰 Keploy: 2024-05-31T22:18:40+05:30 	INFO	Skipping a generated test that failed to increase coverage
🐰 Keploy: 2024-05-31T22:18:40+05:30 	INFO	Running test with the following command: 'npm test'
🐰 Keploy: 2024-05-31T22:18:41+05:30 	INFO	Skipping a generated test that failed
🐰 Keploy: 2024-05-31T22:18:41+05:30 	INFO	Running test with the following command: 'npm test'
🐰 Keploy: 2024-05-31T22:18:42+05:30 	INFO	Skipping a generated test that failed
🐰 Keploy: 2024-05-31T22:18:42+05:30 	INFO	Running test with the following command: 'npm test'
🐰 Keploy: 2024-05-31T22:18:43+05:30 	INFO	Skipping a generated test that failed
🐰 Keploy: 2024-05-31T22:18:43+05:30 	INFO	Running build/test command to generate coverage report: 'npm test'
🐰 Keploy: 2024-05-31T22:18:44+05:30 	INFO	Reached maximum iteration limit without achieving desired coverage. Current Coverage: 58.00%
```

**Comment by charankamarapu** [comment]
Yeah I got 30% - 60%. Its is variable according to chatgpt4.o's mood.

File: pkg/models/ut.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	Reason   string `yaml:"reason"`
+	ExitCode int    `yaml:"exit_code"`
+	Stderr   string `yaml:"stderr"`
+	Stdout   string `yaml:"stdout"`
+	Test     string `yaml:"test"`
+}
+
+type UnitTestsDetails struct {
+	Language                       string     `yaml:"language"`
+	ExistingTestsFunctionSignature string     `yaml:"existing_test_function_signature"`
```
Comment (slayerjain) [inline pkg/models/ut.go:14]: Very long name. Lets try to keep name single word or max 2 words if possible.

File: pkg/service/utgen/coverage.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	FileName string `xml:"filename,attr"`
+	Lines    []Line `xml:"lines>line"`
+}
+
+type Line struct {
+	Number int `xml:"number,attr"`
+	Hits   int `xml:"hits,attr"`
+}
+
+func (cp *CoverageProcessor) ParseCoverageReportCobertura() ([]int, []int, float64, []string, error) {
```
Comment (slayerjain) [inline pkg/service/utgen/coverage.go:88]: This method returns too many things. Please make this an object

File: pkg/models/ut.go (hunk 1 above)
Comment (charankamarapu) [reply pkg/models/ut.go]: refactored

File: pkg/service/utgen/coverage.go (hunk 2 above)
Comment (charankamarapu) [reply pkg/service/utgen/coverage.go]: refactored

File: pkg/service/utgen/gen.go [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+	testFilePath           string
+	codeCoverageReportPath string
+	testCommand            string
+	testCommandDir         string
+	coverageType           string
+	desiredCoverage        float64
+	language               string
+	currentCoverage        float64
+	codeCoverageReport     string
+	relevantLineNumber     int
```
[the remark this answers is not in the document]
Comment (charankamarapu) [reply pkg/service/utgen/gen.go]: refactored

**Comment by charankamarapu** [comment]
Refactored. Please have a convo with @nehagup and finalise on the cmd and flag names too.

**Review by slayerjain — APPROVED** [review]
Looks awesome!

**Pull request merged**